torch/csrc/jit/codegen/cuda/kernel_ir.h

```diff
@@ -813,6 +817,27 @@ class TORCH_CUDA_API BroadcastOp final : public Expr {
   Val* const in_ = nullptr;
 };
 
+class TORCH_CUDA_API TransposeOp final : public Expr {
```

Comment:
why do we need this new node?

Reply:
`kir::TransposeOp` is implemented similarly to `kir::BroadcastOp`. We could avoid introducing this new node, but I thought its own new node would make the overall design more consistent and simpler.

Reply:
What kind of information does the TransposeOp node provide?

A design goal for KIR is to be as close as possible to a normal form (borrowing the definition from databases: https://en.wikipedia.org/wiki/Database_normalization). In particular, this means that there's only one way to represent something - it seems to me that the TransposeOp is just representing a copy in KIR (as opposed to BroadcastOp, which represents a special operation). If it's a copy, we could just use the existing UnaryOp `SET`.

PS. I'm not saying that KIR is in a great shape currently, there's a lot of work to be done to move towards a normal form

Reply:
Yes, it's possible to use the UnaryOp for now. The intent here is that we should create optimized versions of transpose for certain patterns and should call them when applicable. `kir::TransposeOp` represents an entry point for using those optimized versions. I'd imagine `CudaKernelGenerator::visit(kir::TransposeOp*`)` would eventually look similar to `CudaKernelGenerator::visit(kir::BroadcastOp*`)`, which also uses a custom implementation as well as just an assignment.

Reply:
How would that work if the indexing has already "baked in" the permutation? I can see a dedicated node _iff_ it takes care of the permutation (optimized or naive), using the plain indexing as starting point.

Reply:
I thought that would be possible by just manually indexing in a manual CUDA kernel, but for optimized implementations, different loop tiling would be also necessary, so just having arithmetic op for transpose wouldn't be enough. I'll revert the changes in the Kernel IR for now and will revisit it if necessary. Thanks for your feedback!